packages/eslint-config-kodefox/index.js

```diff
@@ -54,7 +54,18 @@ module.exports = {
         noUselessIndex: true,
       },
     ],
-    'import/order': ['warn', { 'newlines-between': 'always' }],
+    'import-helpers/order-imports': [
+      'warn',
+      {
+        newlinesBetween: 'always',
+        groups: [
+          'module',
+          ['parent', 'sibling', 'index'],
+          ['//use[A-Z].*/', '//helper(s?)/'],
+          ['//constant(s?)/', '/type(s?)/'],
+        ],
```

Comment:
Oops, I wrote this like days ago and forgot to submit it.

Reply:
> As of right now, there's no convention about where to put the hooks. So usually I would put them in the same directory as helpers.

That's why I didn't define any directories for hooks and just define `use` as everyone usually put it all over the place, could be anywhere.

> I'm not 100% sure about forcing projects that use this config to match their folder structures based on our personal preferences.

I see. That is a good point. By `their folder structures`, do you mean the user outside our company?

Reply:
> I see. That is a good point. By `their folder structures`, do you mean the user outside our company?

Not necessarily, we have variations in the older projects, some use `lib` and maybe `utils`

Those projects probably can't make use the order unless we add more name variation in the helpers group.

Maybe I worry too much, probably those projects won't use this config anyway.

Reply:
By the way, the grouping detection probably use regex right.

I wonder if the rule will affect the linting duration negatively on a large project.

Reply:
> Not necessarily, we have variations in the older projects, some use `lib` and maybe `utils`

> Those projects probably can't make use the order unless we add more name variation in the helpers group.

Reply:
> I wonder if the rule will affect the linting duration negatively on a large project.

I've run it against one of our old project that has about 400 {ts,tsx} files.

It took **16.59s** to run until it finishes with `1261 problems (88 errors, 1173 warnings)`.

It took **37.11s** to fix those problems until done.

It took **12.43s** re-run it with fewer problems, `387 problems (77 errors, 310 warnings)`.

It took **11.55s** re-run it with `14 problems (2 errors, 12 warnings)`

#### With latest version (0.2.2)

It took **13.75s** to run until it finishes with `796 problems (88 errors, 708 warnings)`.

#### With old version (0.1.0)

It took **9.85s** to run until it finishes with `404 problems (88 errors, 316 warnings)`.

#### Conclusion

The detection by regex indeed makes the lint slower. It's about **2-3** seconds slower.

I don't think it's that slow if we compare it with the latest version only which also using `import/order` but simpler one.

But if we compare it to the old version too, indeed, the cost is high enough as it almost twice the old duration.

I don't know if that cost is acceptable or not. But for me, it's still acceptable as the duration is expected to get longer as the file number growing. As long as it doesn't as long as Jest duration.

#### Bonus

I also tried to run it on exoflex.

It took **4.79s** to run until it finishes with `164 problems (0 errors, 164 warnings)`

It took **4.19s** to run until it finishes with no problem.

Reply:
Sorry I'm so late to this conversation!

> or put this behind option flag.

I like this idea. I've been disabling this on recent projects because it was a bit too strict for my tastes (which is being fixed in this PR I guess). I kinda wonder if we should make this whole package off by default, or maybe just the basic most un-opinionated setting by default. That seems reasonable to me.

> if we're talking about directories, I think the name should always be plural

Correct! Great catch. But if we're being pragmatic, we have to admit that a lot of existing projects probably use the non-plural version, so it might be a safe bet to keep the "s" optional.

But this whole discussion about plurals becomes void if we loosen this up to NOT categorize by folder name, which I'm kinda leaning to that opinion.

@oshimayoan, thanks for doing all the perf measuring, that's very thorough!

In light of that I'm tempted to say we disable all the folder/file name checking by default and just group by built-ins, npm packages, local modules.

One other thought, if there are side effect imports such as `import './polyfills'` then those should probably go first.

Reply:
> One other thought, if there are side effect imports such as import './polyfills' then those should probably go first.

Oops, I forgot about that case. I'll add that case into the rule as well.

> In light of that I'm tempted to say we disable all the folder/file name checking by default and just group by built-ins, npm packages, local modules.

I see. So there will be four if we include the *side effect* imports, right?

Reply:
Speaking of *side effect* imports, what are the patterns for that? Or is it just `polyfills`?

Reply:
Not just polyfill, It could be anything. The way to know it's a side effect import is that it doesn't actually import any named things, it just does puts the path right after the word "import".

> So there will be four if we include the side effect imports, right?

Correct! But I wonder if side effect is a built-in supported category. Not sure about that.

Reply:
> The way to know it's a side effect import is that it doesn't actually import any named things, it just does puts the path right after the word "import".

What if there is a *side effect* that requires an npm package to be imported first?

For example, [this line](https://github.com/kodefox/infra/blob/master/packages/exoflex/src/components/__tests__/Button.test.web.tsx#L4) is require `@testing-library/react` to be imported first.

Reply:
That line should be at the top since we already import `@testing-library/react` inside the file.

Reply:
Seems like the rule already ignoring the order of bare imports.

https://github.com/Tibfib/eslint-plugin-import-helpers/issues/24

Reply:
Yeah, there is support yet from the parent plugin also.
https://github.com/benmosher/eslint-plugin-import/issues/1084

If that the case, I don't think we need this new plugin, right?

Reply:
I think so, unless we want add separate config for additional strictness.

Something like `extends: kodefox/strict`.

Reply:
Maybe it's better if we have a github action that will do fix from the PR.

Just write a comment in the PR, `/format` and then the error will magically go away.

Reply:
I'm not sure there will be any people using that except me 😂

Reply:
> Just write a comment in the PR, `/format` and then import order will be magically fixed

This is new to me!

> I don't think we need this new plugin, right?

> I think so, unless we want add separate config for additional strictness.

Agreed, let's skip the additional strictness for now and if we find a need for it in future we can create a way to turn it on.

Reply:
> This is new to me!
I think this GitHub action doesn't exist yet, we need to make one 😆